apps/prairielearn/assets/scripts/behaviors/react-fragments/registry.tsx

```diff
@@ -0,0 +1,33 @@
+import { type ComponentType } from '@prairielearn/preact-cjs';
+
+import { type DeferredPromise, deferredPromise } from '../../../../src/lib/deferred.js';
+
+type AugmentedDeferredPromise<T> = DeferredPromise<T> & {
+  resolved: boolean;
+};
+
+export class ReactFragmentsRegistry {
+  private fragments: Record<string, AugmentedDeferredPromise<ComponentType<any>>> = {};
+
+  setReactFragment(id: string, component: ComponentType<any>) {
+    if (this.fragments[id] && this.fragments[id].resolved) {
+      throw new Error(`React fragment with id ${id} already resolved`);
+    }
+
+    this.fragments[id] = { ...deferredPromise(), resolved: true };
+    this.fragments[id].resolve(component);
+    this.fragments[id].resolved = true;
```

Comment:
Looks like `resolved` is being set twice?

Reply:
Fixed.